File: src/main/java/no/fintlabs/validation/constraints/IntegrationAndMetadataMatchesValidator.java
package no.fintlabs.validation.constraints;

import no.fintlabs.model.integration.Integration;
import no.fintlabs.model.metadata.IntegrationMetadata;
import no.fintlabs.validation.ConfigurationValidationContext;
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;

import java.util.Objects;

public class IntegrationAndMetadataMatchesValidator implements
        HibernateConstraintValidator<IntegrationAndMetadataMatches, Object> {

    @Override
    public boolean isValid(Object value, HibernateConstraintValidatorContext context) {
        ConfigurationValidationContext configurationValidationContext =
                context.getConstraintValidatorPayload(ConfigurationValidationContext.class);

        Integration integration = configurationValidationContext.getIntegration();
        IntegrationMetadata metadata = configurationValidationContext.getMetadata();

        return Objects.equals(integration.getSourceApplicationId(), metadata.getSourceApplicationId())
                && Objects.equals(integration.getSourceApplicationIntegrationId(), metadata.getSourceApplicationIntegrationId());
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/constraints/IntegrationAndMetadataMatchesValidator.kt
package no.novari.flyt.configuration.validation.constraints

import no.novari.flyt.configuration.validation.ConfigurationValidationContext

class IntegrationAndMetadataMatchesValidator : HibernateConstraintValidator<IntegrationAndMetadataMatches, Any> {
    override fun isValid(
        value: Any,
        hibernateConstraintValidatorContext:
            org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext,
    ): Boolean {
        val configurationValidationContext =
            hibernateConstraintValidatorContext
                .getConstraintValidatorPayload(ConfigurationValidationContext::class.java)

        val integration = configurationValidationContext.integration
        val metadata = configurationValidationContext.metadata

        return integration?.sourceApplicationId == metadata?.sourceApplicationId &&
            integration?.sourceApplicationIntegrationId == metadata?.sourceApplicationIntegrationId
    }
}
